File: core/src/router.rs
//! Policy-driven model routing — pure logic, no I/O. Mirrors the Python router:
//! classify a prompt into (complexity, risk) with a transparent heuristic, then
//! pick a local model from the first matching policy rule.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Level {
    Low,
    Medium,
    High,
}

impl Level {
    fn as_str(self) -> &'static str {
        match self {
            Level::Low => "low",
            Level::Medium => "medium",
            Level::High => "high",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelEntry {
    pub name: String,
    #[serde(default)]
    pub provider: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct When {
    #[serde(default, rename = "taskComplexity")]
    pub task_complexity: Vec<String>,
    #[serde(default, rename = "riskLevel")]
    pub risk_level: Vec<String>,
    #[serde(default, rename = "deviceState")]
    pub device_state: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Rule {
    pub id: String,
    #[serde(default)]
    pub when: When,
    #[serde(rename = "useModel")]
    pub use_model: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Policy {
    #[serde(default)]
    pub models: std::collections::BTreeMap<String, ModelEntry>,
    #[serde(default, rename = "routingRules")]
    pub routing_rules: Vec<Rule>,
}

impl Default for Policy {
    fn default() -> Self {
        let mut models = std::collections::BTreeMap::new();
        models.insert(
            "primary".into(),
            ModelEntry { name: "llama3.2".into(), provider: "ollama".into() },
        );
        Policy {
            models,
            routing_rules: vec![Rule {
                id: "rule_default".into(),
                when: When::default(),
                use_model: "primary".into(),
            }],
        }
    }
}

#[derive(Debug, Clone)]
pub struct RouteDecision {
    pub model: String,
    pub model_key: String,
    pub rule_id: String,
    pub complexity: Level,
    pub risk: Level,
}

const HIGH_RISK: &[&str] = &[
    "delete", "remove", "drop", "deploy", "migrat", "overwrit", "rm ", "sudo",
    "push", "force", "production", "prod", "credential", "password", "secret",
    "payment", "transfer", "wipe", "reset",
];
const HIGH_COMPLEXITY: &[&str] = &[
    "plan", "architect", "design", "analyze", "analyse", "compare", "trade-off",
    "tradeoff", "strateg", "why", "debug", "root cause", "refactor", "step by step",
    "reason",
];

/// Heuristic classification: length + keyword cues. Honest, inspectable.
pub fn classify(prompt: &str) -> (Level, Level) {
    let text = prompt.to_lowercase();
    let words = text.split_whitespace().count();

    let risk = if HIGH_RISK.iter().any(|k| text.contains(k)) {
        Level::High
    } else {
        Level::Low
    };

    let complexity = if words >= 60 || HIGH_COMPLEXITY.iter().any(|k| text.contains(k)) {
        Level::High
    } else if words >= 18 {
        Level::Medium
    } else {
        Level::Low
    };

    (complexity, risk)
}

pub struct Router {
    policy: Policy,
}

impl Router {
    pub fn new(policy: Policy) -> Self {
        Router { policy }
    }

    pub fn from_json(s: &str) -> Self {
        Router::new(serde_json::from_str(s).unwrap_or_default())
    }

    fn rule_matches(when: &When, c: Level, r: Level, device: Option<&str>) -> bool {
        if !when.task_complexity.is_empty()
            && !when.task_complexity.iter().any(|x| x == c.as_str())
        {
            return false;
        }
        if !when.risk_level.is_empty() && !when.risk_level.iter().any(|x| x == r.as_str()) {
            return false;
        }
        if !when.device_state.is_empty() {
            match device {
                Some(d) if when.device_state.iter().any(|x| x == d) => {}
                _ => return false,
            }
        }
        true
    }

    fn resolve(&self, key: &str) -> (String, String) {
        if let Some(e) = self.policy.models.get(key) {
            return (key.to_string(), e.name.clone());
        }
        if let Some((k, e)) = self.policy.models.iter().next() {
            return (k.clone(), e.name.clone());
        }
        ("default".into(), "llama3.2".into())
    }

    pub fn route(&self, prompt: &str, device_state: Option<&str>) -> RouteDecision {
        let (complexity, risk) = classify(prompt);
        let mut key: Option<String> = None;
        let mut rule_id = "none".to_string();
        for rule in &self.policy.routing_rules {
            if Self::rule_matches(&rule.when, complexity, risk, device_state) {
                key = Some(rule.use_model.clone());
                rule_id = rule.id.clone();
                break;
            }
        }
        let key = key.unwrap_or_else(|| {
            self.policy.models.keys().next().cloned().unwrap_or_else(|| "primary".into())
        });
        let (model_key, model) = self.resolve(&key);
        RouteDecision { model, model_key, rule_id, complexity, risk }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn test_policy() -> Policy {
        serde_json::from_str(
            r#"{
              "models": {
                "primary": {"name": "qwen3:14b", "provider": "ollama"},
                "fastFallback": {"name": "qwen3:8b", "provider": "ollama"},
                "deepPlanner": {"name": "deepseek-r1", "provider": "ollama"}
              },
              "routingRules": [
                {"id": "rule_low_power", "when": {"deviceState": ["battery_saver"]}, "useModel": "fastFallback"},
                {"id": "rule_high_risk", "when": {"riskLevel": ["high"]}, "useModel": "deepPlanner"},
                {"id": "rule_deep_path", "when": {"taskComplexity": ["high"]}, "useModel": "deepPlanner"},
                {"id": "rule_fast_path", "when": {"taskComplexity": ["low","medium"], "riskLevel": ["low","medium"]}, "useModel": "primary"},
                {"id": "rule_default", "when": {}, "useModel": "primary"}
              ]
            }"#,
        )
        .unwrap()
    }

    #[test]
    fn simple_goes_fast_path() {
        let r = Router::new(test_policy());
        let d = r.route("what's the date?", None);
        assert_eq!(d.rule_id, "rule_fast_path");
        assert_eq!(d.model, "qwen3:14b");
    }

    #[test]
    fn short_destructive_escalates() {
        let r = Router::new(test_policy());
        let d = r.route("delete the production database", None);
        assert_eq!(d.risk, Level::High);
        assert_eq!(d.rule_id, "rule_high_risk");
        assert_eq!(d.model, "deepseek-r1");
    }

    #[test]
    fn battery_overrides() {
        let r = Router::new(test_policy());
        let d = r.route("what's the date?", Some("battery_saver"));
        assert_eq!(d.rule_id, "rule_low_power");
        assert_eq!(d.model, "qwen3:8b");
    }

    #[test]
    fn never_falls_through() {
        let r = Router::new(test_policy());
        for p in ["hi", "explain why in detail", "rm -rf the server", "summarize my day"] {
            let d = r.route(p, None);
            assert_ne!(d.rule_id, "none");
        }
    }
}

File: cognitive_twin/agent/router.py
"""
Policy-driven model routing — local-first, by rule.

The thesis (shared with local-first agent research like OpenJarvis): keep work on
device and pick the *right local model* for the job instead of sending everything
to one big model — or to the cloud. The routing policy lives in
``policies/model-routing.policy.json`` and is data, not code, so it can change
without touching the loop.

This module:
  1. loads that policy (with a sane built-in default if it's missing),
  2. classifies a request into ``taskComplexity`` + ``riskLevel`` with a small,
     transparent heuristic — no extra model call, stdlib only,
  3. matches the policy's ``routingRules`` in order and returns the chosen model.

The classifier is deliberately a heuristic, not a learned policy: it's honest,
inspectable, and cheap. Swapping in a learned classifier later is a drop-in.
"""

from __future__ import annotations

import json
import os
import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


# --- default policy: used only if policies/model-routing.policy.json is absent.
# Mirrors the shape of the committed policy so behaviour is predictable offline.
_DEFAULT_POLICY: dict[str, Any] = {
    "version": "0-default",
    "models": {
        "primary": {"provider": "ollama", "name": "qwen2.5:7b", "role": "default_reasoning"},
    },
    "routingRules": [
        {"id": "rule_default", "when": {}, "useModel": "primary"},
    ],
    "guardrails": {"allowCloudFallback": False},
}


# Keywords that nudge the heuristic. Kept small and readable on purpose — this is
# a starting signal, not a classifier pretending to be more than it is.
_HIGH_RISK = re.compile(
    r"\b(delete|remove|drop|deploy|migrat|overwrit|rm\s|sudo|push|force|"
    r"production|prod\b|credential|password|secret|payment|transfer|wipe|reset)\b",
    re.IGNORECASE,
)
_HIGH_COMPLEXITY = re.compile(
    r"\b(plan|architect|design|analy[sz]e|compare|trade-?off|strateg|"
    r"why|debug|root cause|refactor|multi-?step|step by step|reason)\b",
    re.IGNORECASE,
)


@dataclass
class RouteDecision:
    """The outcome of routing one request."""
    model: str                       # the resolved Ollama model name
    model_key: str                   # which entry in policy.models was chosen
    rule_id: str                     # which routing rule fired
    task_complexity: str             # low | medium | high
    risk_level: str                  # low | medium | high
    device_state: str | None = None  # e.g. battery_saver, if signalled
    reasons: list[str] = field(default_factory=list)

    def explain(self) -> str:
        bits = [
            f"model={self.model} ({self.model_key})",
            f"rule={self.rule_id}",
            f"complexity={self.task_complexity}",
            f"risk={self.risk_level}",
        ]
        if self.device_state:
            bits.append(f"device={self.device_state}")
        line = "route · " + " · ".join(bits)
        if self.reasons:
            line += "\n        " + "; ".join(self.reasons)
        return line


def load_policy(path: Path | None = None) -> dict[str, Any]:
    """Load the routing policy JSON, falling back to a safe default."""
    candidates = (
        [path] if path is not None else [
            Path(__file__).resolve().parents[2] / "policies" / "model-routing.policy.json",
            Path.cwd() / "policies" / "model-routing.policy.json",
        ]
    )
    for p in candidates:
        try:
            if p and p.is_file():
                return json.loads(p.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            # a broken policy file should degrade to default, not crash the agent
            pass
    return _DEFAULT_POLICY


# Emotional / companionship cues — route these to the warm companion model. Kept
# narrow on purpose: a genuine feeling/check-in turn, NOT anything with a task in
# it (reasoning cues below always win, so "I'm stressed, help me plan X" reasons).
_COMPANION = re.compile(
    r"\b(feel(ing)?|lonely|overwhelmed|anxious|sad|tired|exhausted|stressed|"
    r"miss you|talk to me|just talk|check in|how are you|comfort|vent|"
    r"proud of me|cheer me)\b",
    re.IGNORECASE,
)


def classify(prompt: str) -> tuple[str, str, str, list[str]]:
    """Heuristically estimate (taskComplexity, riskLevel, intent, reasons).

    Transparent by design: length + a few keyword cues. Returns the reasons so
    the decision is inspectable (``--route-explain``).
    """
    reasons: list[str] = []
    text = prompt.strip()
    words = len(text.split())

    # risk
    if _HIGH_RISK.search(text):
        risk = "high"
        reasons.append("matched a high-risk verb (e.g. delete/deploy/secret)")
    else:
        risk = "low"

    # complexity — long prompts or reasoning cues lean higher
    if words >= 60 or _HIGH_COMPLEXITY.search(text):
        complexity = "high"
        reasons.append(
            "reasoning/planning cue or long prompt" if words >= 60 else "matched a reasoning/planning cue"
        )
    elif words >= 18:
        complexity = "medium"
        reasons.append("medium-length request")
    else:
        complexity = "low"

    # intent — companion ONLY when there's an emotional cue AND no reasoning/task
    # cue and it's not a high-complexity/high-risk turn (those need a real brain).
    if (
        _COMPANION.search(text)
        and not _HIGH_COMPLEXITY.search(text)
        and complexity != "high"
        and risk == "low"
    ):
        intent = "companion"
        reasons.append("emotional/check-in cue → companion")
    else:
        intent = "task"

    return complexity, risk, intent, reasons


def _device_state() -> str | None:
    """Optional device hint via env, honouring the policy's low-power rule.
    e.g. CTWIN_DEVICE_STATE=battery_saver | thermal_throttle | normal
    """
    val = (os.environ.get("CTWIN_DEVICE_STATE") or "").strip().lower()
    return val or None


_KNOWN_CONDITIONS = {"taskComplexity", "riskLevel", "deviceState", "intent"}


def _rule_matches(
    when: dict[str, Any], complexity: str, risk: str, device: str | None, intent: str
) -> bool:
    """A rule matches if every condition it specifies is satisfied. An empty
    ``when`` is a catch-all. A rule that references a condition the router does
    NOT understand can never match — otherwise an unknown key would be silently
    ignored and the rule would behave like a catch-all (the bug that sent every
    request to the companion model)."""
    if not when:
        return True
    if any(k not in _KNOWN_CONDITIONS for k in when):
        return False
    if "taskComplexity" in when and complexity not in when["taskComplexity"]:
        return False
    if "riskLevel" in when and risk not in when["riskLevel"]:
        return False
    if "intent" in when and intent not in when["intent"]:
        return False
    if "deviceState" in when:
        if device is None or device not in when["deviceState"]:
            return False
    return True


class Router:
    """Resolves a request to a local model using the routing policy."""

    def __init__(self, policy: dict[str, Any] | None = None) -> None:
        self.policy = policy or load_policy()

    @property
    def allow_cloud_fallback(self) -> bool:
        return bool(self.policy.get("guardrails", {}).get("allowCloudFallback", False))

    def _resolve_model_name(self, model_key: str) -> tuple[str, str]:
        """Map a policy model key (e.g. 'primary') to its concrete name. Falls
        back to the first defined model, then to an installed safe default."""
        models = self.policy.get("models", {})
        entry = models.get(model_key)
        if entry and entry.get("name"):
            return model_key, entry["name"]
        # key not found — use the first model in the policy, else a safe default
        for k, v in models.items():
            if v.get("name"):
                return k, v["name"]
        return "default", "qwen2.5:7b"

    def route(self, prompt: str, *, device_state: str | None = None) -> RouteDecision:
        complexity, risk, intent, reasons = classify(prompt)
        device = device_state if device_state is not None else _device_state()

        chosen_key = None
        fired_rule = "none"
        for rule in self.policy.get("routingRules", []):
            if _rule_matches(rule.get("when", {}), complexity, risk, device, intent):
                chosen_key = rule.get("useModel")
                fired_rule = rule.get("id", "unnamed")
                break

        if chosen_key is None:
            # no rule matched — fall back to the first model, note it
            reasons.append("no routing rule matched; used the first policy model")
            chosen_key = next(iter(self.policy.get("models", {})), "primary")

        model_key, model_name = self._resolve_model_name(chosen_key)
        return RouteDecision(
            model=model_name,
            model_key=model_key,
            rule_id=fired_rule,
            task_complexity=complexity,
            risk_level=risk,
            device_state=device,
            reasons=reasons,
        )
